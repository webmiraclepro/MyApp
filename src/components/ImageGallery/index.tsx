import React, { useState, useEffect, useRef } from 'react'
import {
  StyleSheet,
  Modal,
  FlatList,
  View,
  NativeScrollEvent,
  NativeSyntheticEvent
} from 'react-native'
import { Header, Button } from 'react-native-elements'
import Icon from 'react-native-vector-icons/MaterialIcons'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import ImageViewer, { ImageViewerPropsÂ } from '../ImageViewer'

type ImageProps = {
  imageKeyName?: string
  imageSourceName?: string
}

type Image = Record<string, string | number>

export type ImageGalleryProps = ImageProps & {
  images: Array<Image>,
  isVisible: boolean,
  onClosePress: () => void,
  viewerProps?: Partial<ImageViewerProps>,
  renderFooter?: (bottom: number) => JSX.Element
}

const MAX_SCALE_TO_ENABLE_SCROLL = 1

const ImageGallery = ({
  isVisible,
  images,
  viewerProps = {},
  onClosePress,
  renderFooter,
  imageKeyName = 'key',
  imageSourceName = 'src'
}: ImageGalleryProps) => {
  const listRef = useRef<FlatList<Image>>(null)
  const [index, setIndex] = useState(0)
  const [scrollEnabled, setScrollEnabled] = useState(true)
  const { top, bottom } = useSafeAreaInsets()
  const onScrollEnd = (e: NativeSyntheticEvent<NativeScrollEvent>) => {
    const { contentOffset, layoutMeasurement } = e.nativeEvent
    setIndex(Math.floor(contentOffset.x / layoutMeasurement.width))
  }
  useEffect(() => {
    if (!isVisible) {
      setIndex(0)
    }
  }, [isVisible])

  useEffect(() => {
    setScrollEnabled(true)
  }, [images])
  return (
    <Modal
      visible={isVisible}
      animationType='slide'
      supportedOrientations={['landscape', 'portrait']}
    >
      <Header
        leftComponent={<Button
          accessibilityLabel='imageGalleryClose'
          accessible
          testID='imageGalleryCloseID'
          icon={<Icon
            name='arrow-back-ios'
            color='#000'
            size={26}
          />}
          onPress={onClosePress}
        />}
        centerComponent={{
          text: `${index + 1} of ${images.length}`,
          style: styles.title
        }}
        containerStyle={[styles.header, { height: top + 50, paddingTop: top }]}
      />
      <View style={styles.container}>
        <View style={styles.content}>
          <FlatList
            ref={listRef}
            horizontal
            pagingEnabled
            scrollEnabled={scrollEnabled}
            scrollEventThrottle={200}
            nestedScrollEnabled={scrollEnabled}
            showsHorizontalScrollIndicator={false}
            onMomentumScrollEnd={onScrollEnd}
            data={images}
            renderItem={({ item, index }) => (
              <ImageViewer
                {...viewerProps}
                source={{ uri: String(item[imageSourceName]) }}
                onZoom={(scale) => {
                  const enableScroll = scale <= MAX_SCALE_TO_ENABLE_SCROLL
                  if (!enableScroll && listRef.current) {
                    listRef.current.scrollToIndex({index})
                  }
                  setScrollEnabled(enableScroll)
                }}
              />
            )}
            keyExtractor={(item, index) => String(item[imageKeyName] || index)}
            contentContainerStyle={styles.viewer}
            onStartShouldSetResponder={(e) => {
              return e.nativeEvent.touches.length === 1 && scrollEnabled
            }}
            onMoveShouldSetResponder={(e) => {
              return e.nativeEvent.touches.length === 1 && scrollEnabled
            }}
          />
        </View>
        {renderFooter && (
          <View style={styles.footer}>
            {renderFooter(bottom)}
          </View>
        )}
      </View>
    </Modal>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'flex-end'
  },
  header: {
    justifyContent: 'space-around'
  },
  title: {
    color: '#FFF',
    fontSize: 20,
    textAlignVertical: 'center'
  },
  content: {
    flex: 1,
    flexGrow: 1
  },
  viewer: {
    alignItems: 'center'
  },
  footer: {
    flex: 0
  }
})

export default ImageGallery
