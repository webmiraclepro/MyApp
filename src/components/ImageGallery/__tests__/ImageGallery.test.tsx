import 'react-native'
import React from 'react'
import { mount, shallow } from 'enzyme'

import ImageGallery, { ImageGalleryProps } from '..'

const makeShallowWrapper = (props: ImageGalleryProps) => shallow(
  <ImageGallery {...props} />
)

const makeMountWrapper = (props: ImageGalleryProps) => mount(
  <ImageGallery {...props} />
)

describe('<ImageGallery />', () => {
  const defaultProps: ImageGalleryProps = {
    isVisible: true,
    images: [{ id: 1, src: '123' }, { id: 2, src: '123' }],
    onClosePress: jest.fn()
  }

  it('should render', () => {
    const rendered = makeMountWrapper(defaultProps)
    expect(rendered).toBeTruthy()
  })

  it('renders correctly', () => {
    expect(makeShallowWrapper(defaultProps)).toMatchSnapshot()
  })

  describe('When the close button is pressed', () => {
    it('should call onClosePress function', () => {
      const wrapper = makeMountWrapper(defaultProps)
      wrapper.find({ testID: 'imageGalleryCloseID' }).first().prop('onPress')()

      expect(defaultProps.onClosePress).toHaveBeenCalled()
    })
  })
})
