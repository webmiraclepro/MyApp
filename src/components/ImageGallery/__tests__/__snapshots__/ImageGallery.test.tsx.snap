// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageGallery /> renders correctly 1`] = `
<Component
  animationType="slide"
  hardwareAccelerated={false}
  supportedOrientations={
    Array [
      "landscape",
      "portrait",
    ]
  }
  visible={true}
>
  <ForwardRef(Themed.Header)
    centerComponent={
      Object {
        "style": Object {
          "color": "#FFF",
          "fontSize": 20,
          "textAlignVertical": "center",
        },
        "text": "1 of 2",
      }
    }
    containerStyle={
      Array [
        Object {
          "justifyContent": "space-around",
        },
        Object {
          "height": 50,
          "paddingTop": 0,
        },
      ]
    }
    leftComponent={
      <ForwardRef(Themed.Button)
        accessibilityLabel="imageGalleryClose"
        accessible={true}
        icon={
          <Icon
            allowFontScaling={false}
            color="#000"
            name="arrow-back-ios"
            size={26}
          />
        }
        onPress={[MockFunction]}
        testID="imageGalleryCloseID"
      />
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexGrow": 1,
        }
      }
    >
      <FlatList
        contentContainerStyle={
          Object {
            "alignItems": "center",
          }
        }
        data={
          Array [
            Object {
              "id": 1,
              "src": "123",
            },
            Object {
              "id": 2,
              "src": "123",
            },
          ]
        }
        disableVirtualization={false}
        horizontal={true}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        nestedScrollEnabled={true}
        numColumns={1}
        onEndReachedThreshold={2}
        onMomentumScrollEnd={[Function]}
        onMoveShouldSetResponder={[Function]}
        onStartShouldSetResponder={[Function]}
        pagingEnabled={true}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEnabled={true}
        scrollEventThrottle={200}
        showsHorizontalScrollIndicator={false}
        updateCellsBatchingPeriod={50}
        windowSize={21}
      />
    </View>
  </View>
</Component>
`;
